### PR TITLE
ICU-22520 Replace char arrays with icu::CharString

### PR DESCRIPTION
**Pull request by roubert**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22520
- [X] Required: The PR title must be prefixed with a JIRA Issue number.
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number.
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/common/locdspnm.cpp [hunk 1]
```diff
@@ -575,28 +546,37 @@ LocaleDisplayNamesImpl::localeDisplayName(const Locale& loc,
   UBool hasVariant = uprv_strlen(variant) > 0;
 
   if (dialectHandling == ULDN_DIALECT_NAMES) {
-    char buffer[ULOC_FULLNAME_CAPACITY];
+    UErrorCode status = U_ZERO_ERROR;
```
Comment (markusicu) [inline icu4c/source/common/locdspnm.cpp:549]: Non-ICU-standard indent... we should fix that in a future PR...

File: icu4c/source/common/locdspnm.cpp [hunk 2]
```diff
@@ -575,28 +546,37 @@ LocaleDisplayNamesImpl::localeDisplayName(const Locale& loc,
...
+    CharString buffer;
     do { // loop construct is so we can break early out of search
       if (hasScript && hasCountry) {
-        ncat(buffer, ULOC_FULLNAME_CAPACITY, lang, "_", script, "_", country, (char *)0);
-        localeIdName(buffer, resultName, false);
+        buffer.append(lang, status);
+        buffer.append('_', status);
+        buffer.append(script, status);
+        buffer.append('_', status);
+        buffer.append(country, status);
```
Comment (markusicu) [inline icu4c/source/common/locdspnm.cpp:557]: optional: we can use function chaining
```suggestion
        buffer.append(lang, status).append('_', status).append(script, status).append('_', status).append(country, status);
```

File: icu4c/source/common/locdspnm.cpp [hunk 3]
```diff
@@ -575,28 +546,37 @@ LocaleDisplayNamesImpl::localeDisplayName(const Locale& loc,
...
+        localeIdName(buffer.data(), resultName, false);
```
Comment (markusicu) [inline icu4c/source/common/locdspnm.cpp:558]: Maybe check the status here?

Comment (roubert) [reply icu4c/source/common/locdspnm.cpp:554]: Done.

**Pull request merged**